### PR TITLE
[wip] create maintainers/categories-set.nix

### PR DESCRIPTION
**Pull request by AndersonTorres**
As a proof-of-concept for https://github.com/NixOS/rfcs/pull/146

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 7c6f434c** [comment]
I guess what is needed is also adding category to the meta attribute whitelist?

**Comment by AndersonTorres** [comment]
Isn't this an ofborg-specific config?

**Comment by 7c6f434c** [comment]
There is a list in the manual, and there is `stdenv/generic/check-meta.nix`

**Comment by AndersonTorres** [comment]
Here is the current listing of files inside pkgs directory. They can be useful for filling the categories set.

[listagem.txt](https://github.com/NixOS/nixpkgs/files/12206117/listagem.txt)

**Comment by AndersonTorres** [comment]
@7c6f434c can you please check this again?

File: doc/stdenv/meta.chapter.md [hunk 1]
```diff
@@ -244,6 +244,10 @@ Unfree package that cannot be redistributed. You can build it yourself, but you
 
 This package supplies unfree, redistributable firmware. This is a separate value from `unfree-redistributable` because not everybody cares whether firmware is free.
 
+## Categories {#sec-meta-categories}
+
+The categories of the package. A list of attributes from the attribute set defined in [`nixpkgs/lib/categories-set.nix`](https://github.com/NixOS/nixpkgs/blob/master/lib/categories-set.nix).
```
Comment (7c6f434c) [inline doc/stdenv/meta.chapter.md:249]:
```suggestion
The `meta.categories` attribute contains the categories of the package. A list of attributes from the attribute set defined in [`nixpkgs/lib/categories-set.nix`](https://github.com/NixOS/nixpkgs/blob/master/lib/categories-set.nix).
```

File: maintainers/categories-set.nix [hunk 2]
```diff
@@ -0,0 +1,766 @@
...
+    };
+    ```
+
+    where
+
+    - `handle` is the handle you are going to use in nixpkgs expressions;
+       usually the same as the name but in camelCase
+    - `name` is the category name
+    - `description` is a description of the category
+    - `relatedCategories` is a list of related categories
```
Comment (7c6f434c) [inline maintainers/categories-set.nix:20]: As names or as attribute set references?

Comment (AndersonTorres) [reply maintainers/categories-set.nix:20]: As attrset references.

Comment (7c6f434c) [reply maintainers/categories-set.nix:20]: If there is no example for that, maybe mention this expectation.

Comment (AndersonTorres) [reply maintainers/categories-set.nix:20]: A practical example is not too hard. `audioVideo` is clearly related to `audio` and `video`.

**Comment by AndersonTorres** [comment]
@7c6f434c what do you think of the example of `xa` assembler given above?

**Comment by 7c6f434c** [comment]
I think it works fine as an example, but it feels a bit weird to add specifically «assembler» and not «compiler»

**Comment by AndersonTorres** [comment]
> I think it works fine as an example, but it feels a bit weird to add specifically «assembler» and not «compiler»

Well, the idea is that `meta.categories` allows to be more specific. `xa` is an assembler, not a compiler. 

OK, assemblers are compilers in a broader sense; however, this would be a case for `relatedCategories` (in the example, `development`).

Also, it shows the basic ideas of the data structure.

**Comment by 7c6f434c** [comment]
«assembler → compiler → development» maybe with the transitive «assembler → development» would look natural (with the implication that we don't fill everything, it is just a demonstration), or as you proposed initially «audio → audioVideo» and tag some minor sound processing application like `snd`.

**Comment by AndersonTorres** [comment]
I suggest, then, to never fill the "transitive closure" of every category, in the sense of e.g. `assembler.relatedCategories` is not filled with `compiler.relatedCategories`.

**Comment by AndersonTorres** [comment]
Now it should work as a POC

File: doc/stdenv/meta.chapter.md (hunk 1 above)
Comment (7c6f434c) [reply doc/stdenv/meta.chapter.md:249]: I still think the attribute name should be made explicit like in other entries.

Comment (AndersonTorres) [reply doc/stdenv/meta.chapter.md:249]: Can you explain? I'm not understanding.

Comment (7c6f434c) [reply doc/stdenv/meta.chapter.md:249]: Some sections are titled with _exact_ (including capitalisation) name of the meta attribute. Some sections include the full name in the text near the beginning. Some do both.

The proposed section here does neither, thus my suggestion to add `meta.categories` into the text of the section.

Comment (AndersonTorres) [reply doc/stdenv/meta.chapter.md]: Done.

File: maintainers/categories-set.nix (hunk 2 above)
Comment (AndersonTorres) [reply maintainers/categories-set.nix]: Done, I believe.

**Review by 7c6f434c — APPROVED** [review]
(We will obviously need a bit more process documentation in the manual, but this will need to take into account the choices made in the RFC process)

**Comment by AndersonTorres** [comment]
check-meta conflicted,  closing.

**Pull request closed without merging**